syntax = "proto3";

package hipstershop;

// Product Catalog Service
service ProductCatalogService {
    rpc ListProducts(Empty) returns (ListProductsResponse);
    rpc GetProduct(GetProductRequest) returns (Product);
    rpc SearchProducts(SearchProductsRequest) returns (SearchProductsResponse);
}

// Cart Service
service CartService {
    rpc AddItem(AddItemRequest) returns (Empty);
    rpc GetCart(GetCartRequest) returns (Cart);
    rpc EmptyCart(EmptyCartRequest) returns (Empty);
}

// Currency Service
service CurrencyService {
    rpc GetSupportedCurrencies(Empty) returns (GetSupportedCurrenciesResponse);
    rpc Convert(CurrencyConversionRequest) returns (Money);
}

// Messages
message Empty {}

message Product {
    string id = 1;
    string name = 2;
    string description = 3;
    string picture = 4;
    Money price_usd = 5;
    repeated string categories = 6;
}

message Money {
    string currency_code = 1;
    int64 units = 2;
    int32 nanos = 3;
}

message ListProductsResponse {
    repeated Product products = 1;
}

message GetProductRequest {
    string id = 1;
}

message SearchProductsRequest {
    string query = 1;
}

message SearchProductsResponse {
    repeated Product results = 1;
}

message CartItem {
    string product_id = 1;
    int32 quantity = 2;
}

message Cart {
    string user_id = 1;
    repeated CartItem items = 2;
}

message AddItemRequest {
    string user_id = 1;
    CartItem item = 2;
}

message GetCartRequest {
    string user_id = 1;
}

message EmptyCartRequest {
    string user_id = 1;
}

message GetSupportedCurrenciesResponse {
    repeated string currency_codes = 1;
}

message CurrencyConversionRequest {
    Money from = 1;
    string to_code = 2;
}
